torch/_dynamo/compiled_autograd.py

```diff
@@ -0,0 +1,171 @@
+        self.stack.enter_context(self.proxy_mode)
+        self.stack.enter_context(disable_autocast_cache())
+        self.stack.enter_context(disable_proxy_modes_tracing(enable_current=True))
+        return inputs, sizes
+
+    def proxy_call_hook(self, hook, *args):
+        return self.fx_tracer.create_proxy(
+            "call_function",
+            call_hook,
+            (
+                hook,
+                *[self.to_proxy(x) for x in args],
+            ),
+            {},
+        )
+
+    def tensor_pre_hook(self, inputs, hook_id, i: int):
+        hook = self.hooks_proxy[hook_id]
+        proxy = self.proxy_call_hook(
+            hook,
+            inputs[i],
+        )
+        with disable_proxy_modes_tracing():
+            inputs[i] = inputs[i].clone()
+            self.bind_tensors_to_proxies([inputs[i]], [proxy])
+        return inputs
+
+    def pre_hook(self, inputs, hook_id):
+        hook = self.hooks_proxy[hook_id]
+        proxies = self.proxy_call_hook(
+            hook,
+            inputs,
+        )
+        with disable_proxy_modes_tracing():
+            inputs = [x.clone() for x in inputs]
+            self.bind_tensors_to_proxies(inputs, proxies)
+        return inputs
+
+    def post_hook(self, outputs, inputs, hook_id):
+        hook = self.hooks_proxy[hook_id]
+        proxies = self.proxy_call_hook(
+            hook,
+            outputs,
+            inputs,
+        )
+        with disable_proxy_modes_tracing():
+            outputs = [x.clone() for x in outputs]
+            self.bind_tensors_to_proxies(outputs, proxies)
+        return outputs
+
+    def end_capture(self, outputs):
+        self.stack.close()
+        self.fx_tracer.create_node(
+            "output",
+            "output",
+            (self.fx_tracer.create_arg(self.to_proxy(outputs)),),
+            {},
+        )
+        return self.compiler_fn(
+            GraphModule(self.fx_tracer.root, self.fx_tracer.graph, "CompiledAutograd")
+        )
+
+    def to_proxy(self, t):
+        if t is None:
+            return None
+        if isinstance(t, list):
+            return [self.to_proxy(x) for x in t]
+        if isinstance(t, tuple):
+            return tuple(self.to_proxy(x) for x in t)
+        assert isinstance(t, (torch.Tensor, torch.SymInt))
+        return fetch_tensor_proxy(self.fx_tracer)(t).proxy
+
+    def bind_tensors_to_proxies(self, tensors, proxies):
+        if isinstance(proxies, torch.fx.Proxy):
+            proxies = [proxies[i] for i in range(len(tensors))]
+        assert len(tensors) == len(proxies)
+        assert all(x is not None for x in tensors)
+        assert all(x is not None for x in proxies)
+        track_tensor_tree(tensors, proxies, constant=None, tracer=self.fx_tracer)
+
+
+@contextlib.contextmanager
+def enable(compiler_fn):
+    prior = torch._C._dynamo.compiled_autograd.set_autograd_compiler(
+        functools.partial(AutogradCompilerInstance, compiler_fn)
```

Comment:
Importantly, the AutogradCompilerInstance is reinstantiated every time we do compilation